staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go

```diff
@@ -151,11 +210,103 @@ func WithTypeAndObjectMeta(s *schema.Structural) *schema.Structural {
 		Generic: schema.Generic{Type: "object"},
 		Properties: map[string]schema.Structural{
-			"name": stringType,
+			"name":         stringType,
 			"generateName": stringType,
 		},
 	}
 	result.Properties = props
 
 	return result
-}
\ No newline at end of file
+}
+
+// estimateMinSizeJSON estimates the minimum size in bytes of the given schema when serialized in JSON.
+// minLength/minProperties/minItems are not currently taken into account, so if these limits are set the
+// minimum size might be higher than what estimateMinSizeJSON returns.
+func estimateMinSizeJSON(s *schema.Structural) int64 {
+	switch s.Type {
+	case "boolean":
+		// true
+		return 4
+	case "number", "integer":
+		// 0
+		return 1
+	case "string":
+		if s.ValueValidation != nil {
+			switch s.ValueValidation.Format {
+			case "duration":
+				return minDurationSizeJSON
+			case "date":
+				return dateSizeJSON
+			case "date-time":
+				return minDatetimeSizeJSON
+			}
+		}
+		// ""
+		return 2
+	case "array":
+		// []
+		return 2
+	case "object":
+		// {}
+		objSize := int64(2)
+		// exclude required fields since the request can omit them
+		if s.ValueValidation != nil {
+			for _, propName := range s.ValueValidation.Required {
+				if prop, ok := s.Properties[propName]; ok {
+					// add 4, 2 for quotations around the property name, 1 for the colon, and 1 for a comma
+					objSize += int64(len(propName)) + estimateMinSizeJSON(&prop) + 4
+				}
+			}
+		}
+		return objSize
+	}
+	if s.XIntOrString {
+		// 0
+		return 1
+	}
+	// this code should be unreachable, but return noMaxLength just in case
+	return noMaxLength
+}
+
+// estimateMaxSizeJSON estimates the maximum number of elements that can fit in s considering request size
+// constraints.
+func estimateMaxSizeJSON(s *schema.Structural) int64 {
+	switch s.Type {
+	case "string":
+		if s.ValueValidation != nil {
+			switch s.ValueValidation.Format {
+			case "duration":
+				return maxDurationSizeJSON
+			case "date":
+				return dateSizeJSON
+			case "date-time":
+				return maxDatetimeSizeJSON
+			}
+		}
+		// subtract 2 to account for ""
+		return (maxRequestSizeBytes - 2)
+	case "array":
+		// subtract 2 to account for [ and ]
+		return (maxRequestSizeBytes - 2) / (estimateMinSizeJSON(s.Items) + 1)
+	case "object":
+		if s.AdditionalProperties != nil && s.AdditionalProperties.Structural != nil {
+			// 2 bytes for key + "" + colon + comma + smallest possible value, realistically the actual keys
+			// will all vary in length
+			// TODO(DangerOnTheRanger): is there a way to calculate how many bytes unique keys will need?
+			keyValuePairSize := estimateMinSizeJSON(s.AdditionalProperties.Structural) + 6
+			// subtract 2 to account for { and }
+			return (maxRequestSizeBytes - 2) / keyValuePairSize
+		} else {
+			// this codepath executes in the case of non-map objects,
+			// but regular objects have no concept of maxProperties
+			return noMaxLength
+		}
+	}
+	if s.XIntOrString {
```

Comment:
this block isn't exercised by unit tests

Reply:
Whoops, I'll add a unit test for this branch and resolve this once I do.